Engine/DataFeeds/SubscriptionFrontierTimeProvider.cs

```diff
@@ -0,0 +1,80 @@
+﻿/*
+ * QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
+ * Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+using System;
+
+namespace QuantConnect.Lean.Engine.DataFeeds
+{
+    /// <summary>
+    /// A time provider which updates 'now' time based on the current data emit time of all subscriptions
+    /// </summary>
+    public class SubscriptionFrontierTimeProvider : ITimeProvider
+    {
+        private static readonly long MaxDateTimeTicks = DateTime.MaxValue.Ticks;
+        private DateTime _utcNow;
+        private readonly IDataFeedSubscriptionManager _subscriptionManager;
+
+        /// <summary>
+        /// Creates a new instance of the SubscriptionFrontierTimeProvider
+        /// </summary>
+        /// <param name="utcNow">Initial UTC now time</param>
+        /// <param name="subscriptionManager">Subscription manager. Will be used to obtain current subscriptions</param>
+        public SubscriptionFrontierTimeProvider(DateTime utcNow, IDataFeedSubscriptionManager subscriptionManager)
+        {
+            _utcNow = utcNow;
+            _subscriptionManager = subscriptionManager;
+        }
+
+        /// <summary>
+        /// Gets the current time in UTC
+        /// </summary>
+        /// <returns>The current time in UTC</returns>
+        public DateTime GetUtcNow()
+        {
+            UpdateCurrentTime();
```

Comment:
Have we done any performance profiling? Perhaps for now we can just make sure that `GetUtcNow()` isn't called more often than it needs to be, would be nice if we didn't have to resolve it every time -- not sure how it would work though. For now this is fine, we can optimize the stack when we finish with the architecture refactor.

Reply:
I'll do some performance profiling 👍

Reply:
I've updated the PR description with the performance benchmark data. Doesn't seem to be a performance impact on execution times, for the tested benchmark